### ISSUE TITLE
Does mqtt_cpp supports timeout when trying to connect to the server?

### ISSUE THREAD

**Issue by ineffective** [redboltz/mqtt_cpp#610]
I have a service that tries to connect to one of few servers in a round-robin manner. Usually only one of them is available. Network is pretty fast - usually LAN, which means that if I couldn't connect for a second, I can safely assume that I should try another server. But it seems that currently I have to wait for connection to timeout "naturally", which takes ~30 seconds - is there any way to stop mqtt_cpp from connecting to the server upon server expiration? Maybe there is an example of this?

**Comment by jonesmz** [comment]
You can handle this in a few different ways.

1) Have more than one mqtt_cpp client, each connecting to a different server, in parallel. First one wins, and tells the others to disconnect.
2) Use boost::asio::deadline_timer.

```
boost::asio::deadline_timer timer(io_context);
timer.async_wait(boost::posix_time::seconds(10), [&your_mqtt_cpp_client](std::error_code ec)
{
    if( ! ec)
    {
        // timeout!
        your_mqtt_cpp_client.disconnect();
        // trigger the next server.
    }
});
```

**Comment by ineffective** [comment]
Thanks, timer is basically what I expected, its just that I didn't think that calling `disconnect()` would have such effect - in fact I still don't think so, looking at the code I see that if client is not connected it will still try to connect, as any serious action (including `force_disconnect()`) is only called when socket is, in fact, connected. Should I maybe call `force_disconnect()` by hand? This would call shutdown_from_client() which would, in turn, call `socket.lowest_layer().close(ec)`.

**Comment by jonesmz** [comment]
If the socket isn't connected yet, just destruct your mqtt_cpp::client object.

If the socket is connected, but you haven't yet sent the connect packet (unlikely), it's the same deal.

if you've sent the connect packet, then you should be polite and send the disconnect packet.

**Comment by ineffective** [comment]
Oooooh, so *that's* what you meant when writing about shared pointers, now I understand. It seems that this would work. As a side note, if I were to call `force_disconnect()` by hand, will calling `connect()` later "*resurrect*" client object and it will be able to connect? I don't quite follow the code and I'm possibly missing something.

**Comment by jonesmz** [comment]
> Oooooh, so that's what you meant when writing about shared pointers, now I understand.

I just meant that if you're done with the thing, let the last shared pointer to it go out of scope, and it'll be destructed. Normal shared pointer stuff :-)

> As a side note, if I were to call force_disconnect() by hand, will calling connect() later "resurrect" client object and it will be able to connect?

Off of the top of my head, honestly I'm not sure. You could give it a try, I suppose.

I think though that since the only constructor parameters are the server you're trying to talk to, you should be able to reconnect.

But you won't be able to connect to a different server.

**Comment by ineffective** [comment]
> > Oooooh, so that's what you meant when writing about shared pointers, now I understand.
> 
> I just meant that if you're done with the thing, let the last shared pointer to it go out of scope, and it'll be destructed. Normal shared pointer stuff :-)

Yes, of course, but its just that I failed to see clearly that it would destroy not connected socket - I'd assume that it would be kept forever by io_service/io_context (or whatever is its name).
 
> > As a side note, if I were to call force_disconnect() by hand, will calling connect() later "resurrect" client object and it will be able to connect?
> 
> Off of the top of my head, honestly I'm not sure. You could give it a try, I suppose.
[...]
> But you won't be able to connect to a different server.

Ah, I think you are right. Thanks a lot for your help.

**Comment by Radagan** [comment]
This question is very close to mine. I have an issue with AWS IoT Core that came up in testing, where it waits almost twenty minutes to tell me that my address range for the port is incorrect. 

I'm trying to follow the advice here, but I have the problem that my timer doesn't run until ioc.run() and that's after my call to connect. So, I think I'm missing something important.

Is there an easy way to set a timeout on the call to connect? I'm using the sync client.

Where client_mqtt is:
```
client_mqtt = MQTT_NS::make_tls_sync_client(ioc, broker_url, broker_port, mqtt::protocol_version::v3_1_1);
```

Here's the relevant code:
```
boost::asio::deadline_timer timer(ioc, boost::posix_time::seconds(30));
            timer.async_wait([this](const boost::system::error_code& error) {
                if (error != boost::asio::error::operation_aborted) {
                    spdlog::warn("[MQTT {}] Timeout connecting to broker.", broker_id);
                    client_mqtt->disconnect();
                }
            });
            mqtt::error_code ec;
            client_mqtt->connect(ec);
            timer.cancel();
            if(ec) {
                // error handling removed for posting here...
            }
            spdlog::info("[MQTT {}] Client entering ioc run loop.", broker_id);
            ioc.run();
```
